### PR TITLE
Add lowpass filter when downsampling.

### PR DESCRIPTION
**Pull request by burrows**
before:  [before.wav.gz](https://github.com/greenkeytech/microphone/files/3934341/before.wav.gz)
after:  [after.wav.gz](https://github.com/greenkeytech/microphone/files/3934342/after.wav.gz)

### CODE REVIEW

**Comment by cbrochtrup** [comment]
Looks great! I'll commit the code to make the filter and add some comments about signal processing on Monday.

**Comment by cbrochtrup** [comment]
# Spectrograms of Downsampled Audio
with and without lowpass filtering (anti-aliasing)
## Downsampling with no lowpass filter
Notice the horizontal bands mirroring the bright harmonics ~6kHz. Those are aliased into our spectrogram and should not be there.
image
## Downsampling with lowpass filter
Notice the absence of horizontal bands mirroring the bright harmonics ~6kHz.
image (1)

The spectrograms have vertical bars because they are from old versions of the code without the buffer fix.

File: lowpass_filter_design/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Lowpass Filter Generation
+
+This code generates the lowpass filter to avoid aliasing for audio downsampled to 16kHz. Therefore the cutoff frequency needs to be slightly larger than 8 kHz.
+
+## Quickstart
+
+Generate this filter, or design a new filter
```
Comment (tschady) [inline lowpass_filter_design/README.md]:
```suggestion
Generate this filter, or design a new filter:
```

File: lowpass_filter_design/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Lowpass Filter Generation
+
+This code generates the lowpass filter to avoid aliasing for audio downsampled to 16kHz. Therefore the cutoff frequency needs to be slightly larger than 8 kHz.
```
Comment (tschady) [inline lowpass_filter_design/README.md]: why 8kHz?  a brief explanation would be nice.

File: lowpass_filter_design/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+## Design
+
+The filter designed is a FIR type 1 with with linear phase*. The filter's cutoff frequency is 9.702kHz for a sampling rate of 44.1kHz. The filter order chosen was 41 (or 41 coefficients) because the transition bandh (magnitude response drop off) was deemed acceptable (by Colin) while being a relatively low number for computational efficiency. The cuttoff frqeuency is conservativly high to avoid attenuating energy from 0 to 8 kHz.
+
+The filters frequency response is shown below.
+
+![Lowpass Filter Frequency Response](filter_frequency_response.png)
+
+## Implementation Notes
+If the input sampling rate is greater than 44.1kHz the cuttoff frequency will increase.
```
Comment (tschady) [inline lowpass_filter_design/README.md]:
```suggestion
If the input sampling rate is greater than 44.1kHz the cutoff frequency will increase.
```

File: lowpass_filter_design/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+If redesigning the filter its cuttoff frequency must be greater than or equal to half of the desired sampling rate.
```
Comment (tschady) [inline lowpass_filter_design/README.md]:
```suggestion
If redesigning the filter its cutoff frequency must be greater than or equal to half of the desired sampling rate.
```

File: lowpass_filter_design/generate_filter.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Generates filter coefficients and shows its frequency response
+
+author: Colin Brochtrup
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy import signal
+
+input_fs = 44.1e3
```
Comment (tschady) [inline lowpass_filter_design/generate_filter.py]: Some whitespace here would help.  It's not immediately clear which line this comment belongs to.

```
# comment on var1
var1 = ...

# comment on var2
var2 = ...
```

Comment (cbrochtrup) [reply lowpass_filter_design/generate_filter.py]: Fixed in ddfbcfa5fa241632362249f7bf4b7a5c94ae2d21

File: lowpass_filter_design/README.md (hunk 4 above)
Comment (cbrochtrup) [reply lowpass_filter_design/README.md]: Fixed in ddfbcfa5fa241632362249f7bf4b7a5c94ae2d21

File: lowpass_filter_design/README.md (hunk 3 above)
Comment (cbrochtrup) [reply lowpass_filter_design/README.md]: Fixed in ddfbcfa5fa241632362249f7bf4b7a5c94ae2d21

File: lowpass_filter_design/README.md (hunk 2 above)
Comment (cbrochtrup) [reply lowpass_filter_design/README.md]: Fixed in ddfbcfa5fa241632362249f7bf4b7a5c94ae2d21

File: lowpass_filter_design/README.md (hunk 1 above)
Comment (cbrochtrup) [reply lowpass_filter_design/README.md]: Fixed in ddfbcfa5fa241632362249f7bf4b7a5c94ae2d21

**Pull request merged**